explorer/interpreter/resolve_unformed.h

```diff
@@ -5,11 +5,90 @@
 #ifndef CARBON_EXPLORER_INTERPRETER_RESOLVE_UNFORMED_H_
 #define CARBON_EXPLORER_INTERPRETER_RESOLVE_UNFORMED_H_
 
+#include <optional>
+
 #include "explorer/ast/ast.h"
 #include "explorer/common/nonnull.h"
 
 namespace Carbon {
 
+// Maps AST nodes to flow facts within a function.
+class FlowFacts {
+ public:
+  enum class ActionType {
+    // Used at `VariableDefinition` with initialization.
+    AddInit,
+    // Used at `VariableDefinition` without initialization.
+    AddUninit,
+    // Used at AST nodes that potentially initializes a variable.
+    Form,
+    // Used at AST nodes that uses a variable.
+    Check,
+    // Used in traversing children nodes without an acion to take.
+    None,
+  };
+  // Take action on flow facts based on `ActionType`.
+  auto TakeAction(Nonnull<const AstNode*> node, ActionType action,
+                  SourceLocation source_loc, const std::string& name)
+      -> ErrorOr<Success>;
+
+ private:
+  enum class FormedState {
+    MustBeFormed,
+    MayBeFormed,
+    Unformed,
+  };
+  // Aggregate information about a AstNode being analyzed.
+  struct Fact {
+    FormedState formed_state;
+  };
+
+  auto get_fact(Nonnull<const AstNode*> node) const
+      -> std::optional<Nonnull<const Fact*>> {
+    auto entry = facts_.find(node);
+    if (entry != facts_.end()) {
+      return &entry->second;
+    }
+    return std::nullopt;
+  }
+
+  void add_fact(Nonnull<const AstNode*> node, const FormedState state) {
+    CARBON_CHECK(!get_fact(node).has_value());
```

Comment:
Here and line 71, would `facts_.contains(node)` work? It doesn't seem like you're using the returned value. If that does work, I think `get_fact` would only have one caller in CheckFact, and so may be good to inline.

Reply:
Removed `get_fact` and used facts_.count(node).

Reply:
I meant contains: https://en.cppreference.com/w/cpp/container/map/contains

But I'm realizing that's C++20 so won't work. Probably best to do the `.find` comparison.

Reply:
Yeah that's why I used `.count`. Now they are replaced with `.find`.